src/InvariantTheory/invariant_rings.jl

```diff
@@ -1,130 +1,167 @@
-mutable struct InvariantRing{T1 <: RingElem, T2 <: MatrixElem{T1}}
-   field::Nemo.Field # coefficient field
 
-   group::MatrixGroup{T1, T2}
-   matrices::Vector{T2} # Needs to be set (so far)
+mutable struct InvRing{S, T, U, V, W, X}
+   field::S
+   poly_ring::T
 
-   poly_ring::MPolyRing{T1}
+   group::U
+   action::Vector{V}          # Needs to be set (so far)
+   action_singular::Vector{W} # Needs to be set (so far)
 
-   primary::Vector{<: MPolyElem{T1}}
-   secondary::Vector{<: MPolyElem{T1}}
-   irreducible_secondary::Vector{<: MPolyElem{T1}}
+   modular::Bool
+
+
+   primary::Vector{X}
+   secondary::Vector{X}
+   irreducible_secondary::Vector{X}
+   fundamental::Vector{X}
 
    # Cache some stuff on the Singular side
    # (possibly removed at some point)
    reynolds_singular::Singular.smatrix
    molien_singular::Singular.smatrix
    primary_singular::Singular.smatrix
 
-   function InvariantRing{T1, T2}(matrices::Vector{T2}) where { T1 <: RingElem, T2 <: MatrixElem{T1} }
-      @assert !isempty(matrices)
-      IR = new{T1, T2}()
+   function InvRing(K::S, R::T, G::U, action::Vector{V}) where {S, T, U, V}
+     n = degree(G)
+     R, = PolynomialRing(K, "x" => 1:n)
+     R_sing = singular_ring(R)
+     action_singular = identity.([change_base_ring(R_sing, g) for g in action])
+     X = elem_type(R)
+     W = eltype(action_singular)
+     z = new{S, T, U, V, W, X}()
+     z.field = K
+     z.poly_ring = R
+     z.group = G
+     z.action = action
+     z.action_singular = action_singular
+     z.modular = false
```

Comment:
Isn't this the wrong way around? So, `z.modular` should be `false` if the characteristic is 0 or does not divide the group order because that's the nonmodular case, right?

Reply:
Oops, correct